[Xcode] Enable TAPI_USE_SRCROOT to avoid giving stale headers to tapi

emw-apple opened the pull request:
#### 76096b8d08bca438a88715518bada864c072357a
<pre>
[Xcode] Enable TAPI_USE_SRCROOT to avoid giving stale headers to tapi
<a href="https://bugs.webkit.org/show_bug.cgi?id=250612">https://bugs.webkit.org/show_bug.cgi?id=250612</a>
rdar://103361037

Reviewed by Alexey Proskuryakov.

In recent Xcodes, TAPI_USE_SRCROOT will cause tapi to read headers from
the project&apos;s sources, as reported by Xcode. This is in contrast with
the traditional behavior, where tapi reads *any* header in the
framework&apos;s bundle directory, and is therefore liable to picking up
stale or incorrect headers left from a prior build.

In shipping Xcode, this behavior can be tested by building with
UseSRCROOTSupportForTAPI=1 as an environment variable.

* Configurations/WebKitTargetConditionals.xcconfig: Add target
  conditionals for shipping Xcode versions.

For the following xcconfigs:
1. Enable TAPI_USE_SRCROOT in upcoming Xcodes.
2. When TAPI_USE_SRCROOT is enabled, work around rdar://104432605 by
   passing umbrella headers as pre-includes.

* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/WebKit.xcconfig:
  - WebKit additionally needs a pre-include of config.h, because tapi&apos;s
    SRCROOT mode reads unprocessed headers that depend on config.h being
    imported first (this is rdar://104248994).
  - Use NORMAL_UMBRELLA_FRAMEWORKS_DIR on a line in the diff.
  - Remove reference to NSURLDownloadSPI.h, which is no longer part of
    the target.
* Source/WebKitLegacy/mac/Configurations/WebKitLegacy.xcconfig:

* Tools/Scripts/build-webkit: Set UseSRCROOTSupportForTAPI=YES by
  default where it is supported. This enables SRCROOT mode in EWS.
  Make-based builds are not affected.
* Tools/Scripts/webkitdirs.pm:
(xcodeVersion): Added getter subroutine for $xcodeVersion.
(XcodeOptions): When $ENV{UseSRCROOTSupportForTAPI} is
  truthy, set the TAPI_USE_SRCROOT build setting. This allows us to
  write config logic based on TAPI_USE_SRCROOT that works on older
  Xcodes.

Canonical link: <a href="https://commits.webkit.org/260390@main">https://commits.webkit.org/260390@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2619ed6c47d73874c171d159a8018c6240f065b5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108065 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17133 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40934 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117188 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116518 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111951 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/18618 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8430 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100264 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113832 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13951 "Build was cancelled. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Pull request contains relevant changes; Skipped layout-tests; 2 new passes 11 flakes 16 failures; layout-tests (exception)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97170 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41872 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95858 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28808 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83544 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/97291 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10019 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30156 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/96661 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8113 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10740 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7059 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/30896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16169 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49747 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/105674 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7188 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12321 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26166 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WebGPU/WebGPU/HardwareCapabilities.h:
```diff
@@ -25,6 +25,10 @@
 
 #pragma once
 
+#include "WebGPU.h"
```
Added headers are not explained in the commit message, and I can't say that I understand the additions, especially this one (it smells of some circular dependency).

aproskuryakov commented on Source/WebKit/UIProcess/API/Cocoa/_WKAppHighlightDelegate.h:
```diff
@@ -26,6 +26,7 @@
 #import <Foundation/Foundation.h>
 
 @class _WKAppHighlight;
+@class WKWebView;
```
Adding forward declarations is mentioned in commit message, but I'm not sure if I quite understand this either. Is this about unified build, or something else?

aproskuryakov commented on Tools/Scripts/build-webkit:
```diff
@@ -235,6 +235,13 @@ if (isAppleCocoaWebKit()) {
     }
+
+    if (determineXcodeVersion() lt "15.0" && !defined($ENV{UseSRCROOTSupportForTAPI})) {
+        # The call to XcodeOptions below checks this environment variable and
+        # sets TAPI_USE_SRCROOT for xcconfigs.
+        $ENV{UseSRCROOTSupportForTAPI} = "YES";
```
Confused about why this is needed in `build-webkit`, I expected `webkitdirs.pm` to determine the options in this case. But also, I expected xcconfigs to seethes when appropriate.

emw-apple replied:
In shipping Xcode, the only way to turn on SRCROOT support is via this environment variable or a defaults key. The defaults key is in the `com.apple.dt.XCBuild` domain, _not_ com.apple.dt.Xcode, meaning we can't pass it as an xcodebuild argument. So for testing in EWS, we either need to set it here, or in the buildbot command line.

I can come up with a way to do this in our Makefiles, too, but it'll need to be handled separately.

aproskuryakov commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -192,7 +192,9 @@ REEXPORTED_FRAMEWORK_NAMES[sdk=macosx*] = $(inherited) WebCore;
 // FIXME: Remove -reexport_install_name once rdar://problem/30820233 is fixed.
 OTHER_TAPI_FLAGS = $(OTHER_TAPI_FLAGS_$(WK_COCOA_TOUCH));
-OTHER_TAPI_FLAGS_cocoatouch = $(inherited) -reexport_install_name $(WK_ALTERNATE_WEBKIT_SDK_PATH)$(SYSTEM_LIBRARY_DIR)/PrivateFrameworks/WebKitLegacy.framework/WebKitLegacy -extra-private-header $(SRCROOT)/Platform/ExtraPrivateSymbolsForTAPI.h -extra-public-header $(SRCROOT)/Platform/ExtraPublicSymbolsForTAPI.h -exclude-private-header $(BUILT_PRODUCTS_DIR)/$(PRIVATE_HEADERS_FOLDER_PATH)/NSURLDown ...
+OTHER_TAPI_FLAGS_cocoatouch = $(inherited) -reexport_install_name $(WK_ALTERNATE_WEBKIT_SDK_PATH)$(SYSTEM_LIBRARY_DIR)/PrivateFrameworks/WebKitLegacy.framework/WebKitLegacy -extra-private-header $(SRCROOT)/Platform/ExtraPrivateSymbolsForTAPI.h -extra-public-header $(SRCROOT)/Platform/ExtraPublicSymbolsForTAPI.h $(OTHER_TAPI_FLAGS_SRCROOT_$(TAPI_USE_SRCROOT));
```
As you are touching this line, I think that the path can be changed to use `$(NORMAL_UMBRELLA_FRAMEWORKS_DIR)`.

emw-apple replied:
Ack. Also, I should mention that `-exclude-private-header $(BUILT_PRODUCTS_DIR)/$(PRIVATE_HEADERS_FOLDER_PATH)/NSURLDownloadSPI.h` is removed because it's no longer a private header for this target.

emw-apple committed:
[Xcode] Enable TAPI_USE_SRCROOT to avoid giving stale headers to tapi
https://bugs.webkit.org/show_bug.cgi?id=250612
rdar://103361037

Reviewed by Alexey Proskuryakov.

In recent Xcodes, TAPI_USE_SRCROOT will cause tapi to read headers from
the project's sources, as reported by Xcode. This is in contrast with
the traditional behavior, where tapi reads *any* header in the
framework's bundle directory, and is therefore liable to picking up
stale or incorrect headers left from a prior build.

In shipping Xcode, this behavior can be tested by building with
UseSRCROOTSupportForTAPI=1 as an environment variable.

* Configurations/WebKitTargetConditionals.xcconfig: Add target
  conditionals for shipping Xcode versions.

For the following xcconfigs:
1. Enable TAPI_USE_SRCROOT in upcoming Xcodes.
2. When TAPI_USE_SRCROOT is enabled, work around rdar://104432605 by
   passing umbrella headers as pre-includes.

* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/WebKit.xcconfig:
  - WebKit additionally needs a pre-include of config.h, because tapi's
    SRCROOT mode reads unprocessed headers that depend on config.h being
    imported first (this is rdar://104248994).
  - Use NORMAL_UMBRELLA_FRAMEWORKS_DIR on a line in the diff.
  - Remove reference to NSURLDownloadSPI.h, which is no longer part of
    the target.
* Source/WebKitLegacy/mac/Configurations/WebKitLegacy.xcconfig:

* Tools/Scripts/build-webkit: Set UseSRCROOTSupportForTAPI=YES by
  default where it is supported. This enables SRCROOT mode in EWS.
  Make-based builds are not affected.
* Tools/Scripts/webkitdirs.pm:
(xcodeVersion): Added getter subroutine for $xcodeVersion.
(XcodeOptions): When $ENV{UseSRCROOTSupportForTAPI} is
  truthy, set the TAPI_USE_SRCROOT build setting. This allows us to
  write config logic based on TAPI_USE_SRCROOT that works on older
  Xcodes.

Canonical link: https://commits.webkit.org/260390@main
